src/modm/driver/usb/stusb4500.hpp

```diff
@@ -0,0 +1,160 @@
+		modm::I2cDevice<I2cMaster,4>(address)
+	{};
+
+	/**
+	 * Configures a PDO.
+	 *
+	 * Note: According to the USB standard the first PDO (1) always has to set
+	 *  voltage to 5000 [mV].
+	 *
+	 * @param	pdoNumber Which PDO to configure, range 1 to 3.
+	 * @param	voltage	  in mV
+	 * @param	current	  in mA
+	 */
+	inline modm::ResumableResult<bool>
+	configurePdo(uint8_t pdoNumber, uint32_t voltage, uint32_t current)
+	{
+		RF_BEGIN();
+
+		if((pdoNumber < 1) || (pdoNumber > 3))
+			RF_RETURN(false);
+
+		RF_CALL(readRegister<4>(DpmSnkPdos.at(pdoNumber - 1), buffer2.data()));
+
+		current /= 10;
+		voltage /= 50;
+
+		buffer2.at(0) = current;
+		buffer2.at(1) = ((current >> 8) & 0b11) | ((voltage & 0b11'1111 ) << 2);
+		buffer2.at(2) = (buffer2.at(2) & 0b1111'0000) | ((voltage >> 6) & 0b1111);
+
+		RF_END_RETURN_CALL(updateRegister<4>(DpmSnkPdos.at(pdoNumber - 1), buffer2.data()));
+	}
+
+	/**
+	 * Sets the valid PDO.
+	 *
+	 * @param	pdoNumber Which PDO to set valid, range 1 to 3.
+	 */
+	modm::ResumableResult<bool>
+	setValidPdo(uint8_t pdoNumber)
+	{
+		RF_BEGIN();
+
+		if((pdoNumber < 1) || (pdoNumber > 3))
+			RF_RETURN(false);
+
+		RF_END_RETURN_CALL(updateRegister<1>(Register::DpmPdoNumb, &pdoNumber));
+	}
+
+	/**
+	 * Retrieve RDO status from STUSB4500 chip.
+	 *
+	 * @return	RdoRegStatusData object. Typically only
+	 *  RdoRegStatusData::MaxCurrent [mA] is of interest.
+	 */
+	inline modm::ResumableResult<RdoRegStatusData>
+	getRdoRegStatus()
+	{
+		RF_BEGIN();
+
+		RF_CALL(readRegister<4>(Register::RdoRegStatus, buffer2.data()));
+
+		// MaxCurrent: Bits 9..0; OperatingCurrent: Bits 19..10; ObjectPos: Bits 30..28
+		rdoRegStatusData.MaxCurrent = buffer2.at(0) | ((buffer2.at(1) & 0b11) << 8);
+		rdoRegStatusData.MaxCurrent *= 10;
+		rdoRegStatusData.OperatingCurrent = ((buffer2.at(1) & 0b1111'1100) >> 2) | ((buffer2.at(2) & 0b1111) << 6);
+		rdoRegStatusData.OperatingCurrent *= 10;
+		rdoRegStatusData.ObjectPos = (buffer2.at(3) & 0b0110'0000) >> 5;
+
+		RF_END_RETURN(rdoRegStatusData);
+	}
+
+private:
+	template<uint32_t length>
+	modm::ResumableResult<bool>
+	readRegister(Register reg, uint8_t* output)
+	{
+		RF_BEGIN();
+		buffer[0] = uint8_t(reg);
+		this->transaction.configureWriteRead(buffer.data(), 1, output, length);
+		RF_END_RETURN_CALL( this->runTransaction() );
+	}
+
+	template<uint32_t length>
+	modm::ResumableResult<bool>
+	updateRegister(Register reg, uint8_t* data)
+	{
+		static_assert(length <= (std::tuple_size<decltype(buffer)>::value - 1), "buffer too small");
+
+		RF_BEGIN();
+		buffer.at(0) = uint8_t(reg);
+		std::memcpy(&buffer.at(1), data, length);
+		this->transaction.configureWriteRead(buffer.data(), length + 1, nullptr, 0);
+		RF_END_RETURN_CALL( this->runTransaction() );
+	}
+
+private:
+	static constexpr std::array<Register, 3> DpmSnkPdos {Register::DpmSnkPdo1, Register::DpmSnkPdo2, Register::DpmSnkPdo3};
+	std::array<uint8_t, 5> buffer;
+	std::array<uint8_t, 4> buffer2;
```

Comment:
Why are we using `std::array` for such small buffers? The [`.at` method](https://en.cppreference.com/w/cpp/container/array/at) does bounds checking (and throws an exception), but we only ever call it with fixed numbers. It'll probably be optimized away, but it's still kinda weird?

Reply:
Also: buffer2 is unnecessary?